File: dotnet/CardsProject/Services/Tournaments/TournamentRegistrationService.cs
using CardsProject.Domain.Tournaments;
using CardsProject.Infrastructure;

namespace CardsProject.Services.Tournaments;

public class TournamentRegistrationService
{
    private readonly AppDbContext _db;

    public TournamentRegistrationService(AppDbContext db) => _db = db;

    public async System.Threading.Tasks.Task<TournamentRegistration> CreateAsync(TournamentRegistration entity)
    {
        _db.TournamentRegistrations.Add(entity);
        await _db.SaveChangesAsync();
        return entity;
    }

    public async System.Threading.Tasks.Task<TournamentRegistration> UpdateAsync(TournamentRegistration entity)
    {
        _db.TournamentRegistrations.Update(entity);
        await _db.SaveChangesAsync();
        return entity;
    }

    public async System.Threading.Tasks.Task<bool> WithdrawAsync(int id)
    {
        var entity = await _db.TournamentRegistrations.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("TournamentRegistration not found: " + id);
        entity.Withdraw();
        await _db.SaveChangesAsync();
        return true;
    }
    public async System.Threading.Tasks.Task<bool> DisqualifyAsync(int id, string reason)
    {
        var entity = await _db.TournamentRegistrations.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("TournamentRegistration not found: " + id);
        entity.Disqualify(reason);
        await _db.SaveChangesAsync();
        return true;
    }
    public async System.Threading.Tasks.Task<bool> PromoteFromWaitlistAsync(int id)
    {
        var entity = await _db.TournamentRegistrations.FindAsync(id);
        if (entity is null) throw new KeyNotFoundException("TournamentRegistration not found: " + id);
        entity.PromoteFromWaitlist();
        await _db.SaveChangesAsync();
        return true;
    }
}

File: laravel/app/Services/Tournaments/TournamentRegistrationService.php
<?php

namespace App\Services\Tournaments;

use App\Models\Tournaments\TournamentRegistration;

class TournamentRegistrationService
{
    public function create(array $data): TournamentRegistration
    {
        throw new \LogicException('Not implemented');
    }

    public function update(TournamentRegistration $tournamentRegistration, array $data): TournamentRegistration
    {
        throw new \LogicException('Not implemented');
    }
    public function withdraw(int $id): void
    {
        $tournamentRegistration = TournamentRegistration::findOrFail($id);
        $tournamentRegistration->withdraw();
        $tournamentRegistration->save();
    }

    public function disqualify(int $id, $reason): void
    {
        $tournamentRegistration = TournamentRegistration::findOrFail($id);
        $tournamentRegistration->disqualify($reason);
        $tournamentRegistration->save();
    }

    public function promoteFromWaitlist(int $id): void
    {
        $tournamentRegistration = TournamentRegistration::findOrFail($id);
        $tournamentRegistration->promoteFromWaitlist();
        $tournamentRegistration->save();
    }
}
